library/std/src/sys/windows/args.rs

```diff
@@ -28,129 +35,121 @@ pub fn args() -> Args {
 /// Implements the Windows command-line argument parsing algorithm.
 ///
 /// Microsoft's documentation for the Windows CLI argument format can be found at
-/// <https://docs.microsoft.com/en-us/previous-versions//17w5ykft(v=vs.85)>.
+/// <https://docs.microsoft.com/en-us/cpp/cpp/main-function-command-line-args?view=msvc-160#parsing-c-command-line-arguments>
 ///
-/// Windows includes a function to do this in shell32.dll,
-/// but linking with that DLL causes the process to be registered as a GUI application.
+/// A more in-depth explanation is here:
+/// <https://daviddeley.com/autohotkey/parameters/parameters.htm#WIN>
+///
+/// Windows includes a function to do command line parsing in shell32.dll.
+/// However, this is not use for two reasons:
+///
+/// 1. Linking with that DLL causes the process to be registered as a GUI application.
 /// GUI applications add a bunch of overhead, even if no windows are drawn. See
 /// <https://randomascii.wordpress.com/2018/12/03/a-not-called-function-can-cause-a-5x-slowdown/>.
 ///
-/// This function was tested for equivalence to the shell32.dll implementation in
-/// Windows 10 Pro v1803, using an exhaustive test suite available at
-/// <https://gist.github.com/notriddle/dde431930c392e428055b2dc22e638f5> or
-/// <https://paste.gg/p/anonymous/47d6ed5f5bd549168b1c69c799825223>.
-unsafe fn parse_lp_cmd_line<F: Fn() -> OsString>(
-    lp_cmd_line: *const u16,
+/// 2. It does not follow the modern C/C++ argv rules outlined in the first two links above.
+///
+/// This function was tested for equivalence to the C/C++ parsing rules using an
+/// extensive test suite available at
+/// <https://github.com/ChrisDenton/winarg/tree/std>.
+fn parse_lp_cmd_line<'a, F: Fn() -> OsString>(
+    lp_cmd_line: Option<WStrUnits<'a>>,
     exe_name: F,
 ) -> Vec<OsString> {
-    const BACKSLASH: u16 = '\\' as u16;
-    const QUOTE: u16 = '"' as u16;
-    const TAB: u16 = '\t' as u16;
-    const SPACE: u16 = ' ' as u16;
+    // SAFETY: These are all non-zero ASCII characters.
+    const BACKSLASH: NonZeroU16 = unsafe { NonZeroU16::new_unchecked(b'\\' as u16) };
+    const QUOTE: NonZeroU16 = unsafe { NonZeroU16::new_unchecked(b'"' as u16) };
+    const TAB: NonZeroU16 = unsafe { NonZeroU16::new_unchecked(b'\t' as u16) };
+    const SPACE: NonZeroU16 = unsafe { NonZeroU16::new_unchecked(b' ' as u16) };
```

Comment:
Random suggestion from an outsider, using `NonZeroU16::new` with `unwrap` seems better than using the unsafe function.
```suggestion
    const BACKSLASH: NonZeroU16 = NonZeroU16::new(b'\\' as u16).unwrap();
    const QUOTE: NonZeroU16 = NonZeroU16::new(b'"' as u16).unwrap();
    const TAB: NonZeroU16 = NonZeroU16::new(b'\t' as u16).unwrap();
    const SPACE: NonZeroU16 = NonZeroU16::new(b' ' as u16).unwrap();
```

Reply:
@Urgau I definitely agree! But the `const` context means I'd need to use the unstable `#![feature(const_option)]` which doesn't appear to be enabled for the standard library yet.

Reply:
Ah, sorry I only check `core`, but because the feature is enable in `core`, it can also probably be enabled in `std`. If you want to enable it, should add in the [lib.rs](https://github.com/rust-lang/rust/blob/master/library/std/src/lib.rs) of `std`.

Reply:
That sounds reasonable, thanks! I've added it. If anyone has reason not to want this feature in `std` I can always revert the commit.